sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -84,84 +129,164 @@ case class SortMergeJoin(
+                bufferedPosition = 0
+                if (joinType != FullOuter || secondStreamedElement == null) {
+                  fetchStreamed()
+                  if (streamedElement == null || keyOrdering.compare(streamedKey, matchKey) != 0) {
+                    stop = false
+                    bufferedMatches = null
+                  }
+                } else {
+                  // in FullOuter join and the first time we finish the match buffer,
+                  // we still want to generate all rows with streamed null row and buffered
+                  // rows that match the join key but not the conditions.
+                  streamedElement = secondStreamedElement
+                  bufferedMatches = secondBufferedMatches
+                  secondStreamedElement = null
+                  secondBufferedMatches = null
+                }
               }
+              joinedRow
             }
-            joinedRow
           } else {
             // no more result
             throw new NoSuchElementException
           }
         }
 
-        private def fetchLeft() = {
-          if (leftIter.hasNext) {
-            leftElement = leftIter.next()
-            leftKey = leftKeyGenerator(leftElement)
+        private def smartJoinRow(streamedRow: Row, bufferedRow: Row): Row = joinType match {
+          case RightOuter => joinRow(bufferedRow, streamedRow)
+          case _ => joinRow(streamedRow, bufferedRow)
+        }
+
+        private def fetchStreamed() = {
+          if (streamedIter.hasNext) {
+            streamedElement = streamedIter.next()
+            streamedKey = streamedKeyGenerator(streamedElement)
           } else {
-            leftElement = null
+            streamedElement = null
           }
         }
 
-        private def fetchRight() = {
-          if (rightIter.hasNext) {
-            rightElement = rightIter.next()
-            rightKey = rightKeyGenerator(rightElement)
+        private def fetchBuffered() = {
+          if (bufferedIter.hasNext) {
+            bufferedElement = bufferedIter.next()
+            bufferedKey = bufferedKeyGenerator(bufferedElement)
           } else {
-            rightElement = null
+            bufferedElement = null
           }
         }
 
         private def initialize() = {
-          fetchLeft()
-          fetchRight()
+          fetchStreamed()
+          fetchBuffered()
         }
 
         /**
          * Searches the right iterator for the next rows that have matches in left side, and store
          * them in a buffer.
+         * When this is not a Inner join, we will also return true when we get a row with no match
+         * on the other side. This search will jump out every time from the same position until
+         * `next()` is called.
          *
          * @return true if the search is successful, and false if the right iterator runs out of
          *         tuples.
          */
         private def nextMatchingPair(): Boolean = {
```

Comment:
Reading through the old version of the code, one tricky thing that stood out to me was the fact that if a consumer of this iterator calls `hasNext()` followed by `next()`, then we end up calling `nextMatchingPair()` two times in a row.  It might be nice to add a comment here to explain why this is safe / correct.